### PR TITLE
Paper: Securing Your Collaborative Jupyter Notebooks in theCloud using Container and Load Balancing Services

### PR DESCRIPTION
**Pull request by summerswallow-whi**
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

### CODE REVIEW

**Comment by emmjab** [comment]
Hi @summerswallow-whi ! Introducing myself here, as I'll be reviewing your paper!

**Comment by dhavide** [comment]
Hello @summerswallow-whi, this is @dhavide and I'll be one of the reviewers.
I will read through the manuscript carefully over the next few days and offer some feedback before June 18th.

**Comment by gcdeshpande** [comment]
Hello @summerswallow-whi. I am one of the reviewers assigned to your paper.
At the outset I would like to congratulate you for writing a good paper. I will write my comments in couple of hours

**Comment by gcdeshpande** [comment]
1. Please number sections and subsections in the paper. 
2. Check for grammatical and typographical mistakes. I found a few.
3. I think in some cases the length of the sentences can be reduced.
4. I feel the literature review section can be added. This helps the reader to understand the current status of security in Jypyter notebooks and others. Also, you can add your contributions to improve security.
5. Please number and add a short description to code listings
6. I would like to know the details of the experiments done to prove security implementation. In case they are available kindly share the details. Any metrics used to measure security parameters?

**Comment by dhavide** [comment]
With reference to the [recommended SciPy reviewing criteria](https://github.com/scipy-conference/scipy_proceedings/blob/2019/review_criteria.md), there are two concerns of the current draft (both of which are, IMHO, resolvable) as a submission to SciPy 2020:

1. the results are not easy to verify in the sense of reproducibility; and
2. the connections to the scientific Python audience are not obvious.

The first point is probably an unfair criterion in that the paper describes a fairly significant data engineering problem and its pragmatic solution. The technical content seems sound but it is tricky to verify. Verification of the results would require acquring suitable compute resources from Amazon and deploying them as suggested. The handful of code snippets illustrate some of the generality of the framework quite well, but it is, of course, difficult to get a sense of what's going on from that alone. One other hindrance is that the link to [https://github.com/Westhealth/scipy2020/cloud_infrastructure](https://github.com/Westhealth/scipy2020/cloud_infrastructure) at the end of the manuscript did not work for me (404 error); if access to this code is required to understand the manuscript, please correct this.

The second point, to my mind, is more important but also easier to fix. Full disclosure: data engineering is not my strength as compared to scientific computing, so this slightly negative opinion reflects my own interests and biases. My suspicion is that the SciPy audience, while consisting of folks from industry and academia alike, does not have significant data engineering expertise. Although the authors do an excellent job explaining the many technical challenges (e.g., of encryption at-rest vs. in-transit), I believe that the typical SciPy audience member will not be drawn in immediately. My suggestion is to find a way to connect the discussion more directly to a scientific computing audience. This can probably be done with a sentence or two added in the introduction & conclusion. It would also be helpful to add a few explanatory sentences in the more technical content, i.e., why this particular problem is important/difficult, why your solution works well, and some cursory evidence to support that assertion. As an outsider, a little more context would help.

The quality of the written English is overall quite good. There are a handful of places that can be improved easily (minor spelling errors, brief shifts into passive voice, change in verb tense, a few unexplained acronyms, etc.); I'll flag these explicitly on a final reviewing pass. These are not significant detractors. There is a reasonable attempt to explain acronyms and such (although, given the subject matter, it's impossible to avoid some fairly arcane jargon).

All this said, I think this is an interesting manuscript that is appropriate for publication in SciPy 2020's proceedings. I am happy to clarify these comments and to work with the authors to improve the manuscript.

**Comment by summerswallow-whi** [comment]
@gcdeshpande
> 1. Please number sections and subsections in the paper.
> 2. Check for grammatical and typographical mistakes. I found a few.
> 3. I think in some cases the length of the sentences can be reduced.
> 4. I feel the literature review section can be added. This helps the reader to understand the current status of security in Jypyter notebooks and others. Also, you can add your contributions to improve security.
> 5. Please number and add a short description to code listings
> 6. I would like to know the details of the experiments done to prove security implementation. In case they are available kindly share the details. Any metrics used to measure security parameters?

1) I looked at the example and past years proceedings, doesn't look like the standard Scipy format includes section numbering, but if you think they should be there we should probably get clarification.
2) I'll review it again 
3) I'll review and see what makes sense to break up.
4) Literature is scant in this area but I will breakout a section into `background` which will discuss the state of the art a little more.
5) I can once I figure out how. (If you know how to do this in the scipy rst please share but I will try to figure it out)
6) We actually hired third party penetration tester. I'll expand the explanation of this into a paragraph.

**Comment by summerswallow-whi** [comment]
> With reference to the [recommended SciPy reviewing criteria](https://github.com/scipy-conference/scipy_proceedings/blob/2019/review_criteria.md), there are two concerns of the current draft (both of which are, IMHO, resolvable) as a submission to SciPy 2020:
> 
> 1. the results are not easy to verify in the sense of reproducibility; and
> 2. the connections to the scientific Python audience are not obvious.
> 
> The first point is probably an unfair criterion in that the paper describes a fairly significant data engineering problem and its pragmatic solution. The technical content seems sound but it is tricky to verify. Verification of the results would require acquring suitable compute resources from Amazon and deploying them as suggested. The handful of code snippets illustrate some of the generality of the framework quite well, but it is, of course, difficult to get a sense of what's going on from that alone. One other hindrance is that the link to https://github.com/Westhealth/scipy2020/cloud_infrastructure at the end of the manuscript did not work for me (404 error); if access to this code is required to understand the manuscript, please correct this.
> 
> The second point, to my mind, is more important but also easier to fix. Full disclosure: data engineering is not my strength as compared to scientific computing, so this slightly negative opinion reflects my own interests and biases. My suspicion is that the SciPy audience, while consisting of folks from industry and academia alike, does not have significant data engineering expertise. Although the authors do an excellent job explaining the many technical challenges (e.g., of encryption at-rest vs. in-transit), I believe that the typical SciPy audience member will not be drawn in immediately. My suggestion is to find a way to connect the discussion more directly to a scientific computing audience. This can probably be done with a sentence or two added in the introduction & conclusion. It would also be helpful to add a few explanatory sentences in the more technical content, i.e., why this particular problem is important/difficult, why your solution works well, and some cursory evidence to support that assertion. As an outsider, a little more context would help.
> 
> The quality of the written English is overall quite good. There are a handful of places that can be improved easily (minor spelling errors, brief shifts into passive voice, change in verb tense, a few unexplained acronyms, etc.); I'll flag these explicitly on a final reviewing pass. These are not significant detractors. There is a reasonable attempt to explain acronyms and such (although, given the subject matter, it's impossible to avoid some fairly arcane jargon).
> 
> All this said, I think this is an interesting manuscript that is appropriate for publication in SciPy 2020's proceedings. I am happy to clarify these comments and to work with the authors to improve the manuscript.

On point one: Obviously there is not space in the manuscript to include enough code to make it reproducible. However, the intent is that code be placed in the link by the time the proceedings is made available. The link is at present broken, primarily because the code really isn't properly commented and I have to have cleared what should be made public, but if it helps I will expedite the matter. The code provided in the link should not be needed to understand the paper, but is available to make it easier to implement if someone wishes. I'm working now to get some code released.

Second point: I will endeavor to address it. Certainly a sentence or two I can add. I'll work on this soon

**Comment by summerswallow-whi** [comment]
@gcdeshpande. I have looked around to see if there were an easy way to  add numbering and a caption to the listings. I know this can be done in latex, but given the constraints imposed by using rst in the scipy environment. I can't find a way to do this. If you know how, please share.

Or do you mean add line numbers? I felt line numbers would be too busy, but I can add them if that is what you mean.

**Comment by summerswallow-whi** [comment]
I've revised the document and I'm sure it will take some more iterations. But since this is supposed to be an iterative process, I figured best to push changes.

First, I did not proofread for grammatical errors though I did fix some spelling mistakes as I encountered them. as @dhavide  said we can do that as we finalize.

@gcdeshpande Your points 4 & 6 are addressed.  Points 1 & 5 are pending your clarification and 2 & 3 I'll address as we finalize as no doubt typos and spelling mistakes will creep in in the review process. I will highlight my changes directly in the documents.

File: papers/72_lu/72_lu.rst [hunk 1]
```diff
@@ -46,14 +46,12 @@ Introduction
...
 (AWS) though many architectures and methods described here are applicable to
-other cloud platforms. 
-
-Project Jupyter includes Jupyter Hub, which provides a proxy and container
-management. In particular the Zero to Jupyter Hub with Kubernetes project,
-:cite:`z2k8s` provides a framework to implement Jupyter Hub on a Kubernetes
-platform but with many significant security drawbacks. Kubernetes is
-notoriously difficult to secure and has many vulnerabilities that are not
-addressed by default.
+other cloud platforms. As Jupyter is the mainstay of scientific
```
Comment (summerswallow-whi) [inline papers/72_lu/72_lu.rst:49]: @dhavide This revision in the paragraph endeavors to address your point about connecting the a broader readership. I would appreciate any suggestions that might improve this.

File: papers/72_lu/72_lu.rst [hunk 2]
```diff
@@ -78,6 +76,36 @@ Our specific implementation satisfies privacy and security concerns and can
 serve as a starting point to develop customized solutions for related use
 cases.
 
+Background
```
Comment (summerswallow-whi) [inline papers/72_lu/72_lu.rst:79]: @gcdeshpande I broke out  a background section. As I stated there isn't much literature on security of Jupyter (which is part of the problem). More a prevailing knowledge that it is not particularly secure.  Hopefully, this fills your request. If not, I would appreciate suggestions.

File: papers/72_lu/72_lu.rst [hunk 3]
```diff
@@ -252,7 +280,7 @@ The solution to the last requirement is for our provisioner to write an ``.id``
 Jupyter 
 +++++++
 
-Unfortunately, unlike JupyterHub, Jupyter notebook/lab do not come with a pluggable authentication module. In order to implement validation, the source file :code:`login.py` must be modified. This file is usually located in the :code:`notebook/auth/` directory in your :code:`site-packages` or ``dist-packages`` directory. Since Jupyter notebook and JupyterLab are not truly separate applications (in fact they are interchangeable using the path ``/tree`` or ``/lab``), the same ``login.py`` file facilitates authentication for both. If you build using a standard docker image such as :code:`jupyter/base-notebook` or any of its derivative notebooks, this directory would be :code:`/opt/conda/lib/python3.x/site-packages` directory. Please note that the specific python version may vary dependent on which version of the docker container is used and whether subsequent additional install modules might force a rollback of python versions.
```
Comment (summerswallow-whi) [inline papers/72_lu/72_lu.rst]: @dhavide This is one place I did find the problem particularly challenge. If you spot other places, I am certainly receptive to adding more reference.

File: papers/72_lu/72_lu.rst [hunk 4]
```diff
@@ -606,17 +634,25 @@ Conclusion
 ----------
 
 Presented here is a secure, collaborative infrastructure for deploying a cloud
-computation resources. The primary purpose of our infrastructure is to provide
-Jupyter in this environment though due to the preference of some of our users
-RStudio and other tools are include. Our Data Science and infrastructure team
+computation resources vital to our organization for scientific
```
Comment (summerswallow-whi) [inline papers/72_lu/72_lu.rst:637]: @dhavide Another attempt to tie the paper to a more generic audience. Again suggestions welcome

File: papers/72_lu/72_lu.rst [hunk 5]
```diff
@@ -606,17 +634,25 @@ Conclusion
...
+RStudio. Our Data Science and infrastructure team 
 is small so building a compliant infrastructure that requires little
 maintenance is paramount. Equally important is to safeguard against opening
 vulnerabilities due to misconfigurations. By following the suggestions
 presented here, misconfigurations err on the side of loss of functionality
 rather than introducing vulnerabilites.
 
 The architecture presented here was successful in a recently performed
-penetration test. While the recommendations and architecture shown here rely
+penetration test. We hired a third party company that specializes in
```
Comment (summerswallow-whi) [inline papers/72_lu/72_lu.rst:649]: @gcdeshpande Regarding your comment about evaluating the security of the system. I am permitted to divulge more details, so here it. For us the best way is to hire a professional cyber risk assessment company and see if they can break into our infrastructure. I tend not to put vendor names in the paper when possible.

**Comment by deniederhut** [comment]
Hi everyone! I'd like to weigh in on a few topics that have come up so far.

> I looked at the example and past years proceedings, doesn't look like the standard Scipy format includes section numbering

Yes, in the past we have not numbered manuscript sections, and in general ask that authors not alter the latex stylesheet so that we can keep a consistent format across papers.

> The first point is probably an unfair criterion in that the paper describes a fairly significant data engineering problem and its pragmatic solution

I haven't had a chance to look through the paper yet, but the 'reproducibility' thing is problematic for two reasons:

1. Many authors submit papers from government offices, which are unable to publish code for security reasons; and,
2. Many authors submit papers about implementations (as opposed to experiments), which are not reproducible in the classic sense of recovering a given output.

With that context, please take this review criterion to mean that we encourage openness where possible, but we do not want to require it (and exclude amazing work done by, for example, the US Airforce).

> Literature is scant in this area

It might be helpful to reach out to @choldgraf or one of the other members of the Jupyter core team for some feedback here. They are very nice and helpful, but like all nice and helpful people, they can get a bit oversubscribed.

**Comment by summerswallow-whi** [comment]
Thanks for your input.
> 
> > I looked at the example and past years proceedings, doesn't look like the standard Scipy format includes section numbering
> 
> Yes, in the past we have not numbered manuscript sections, and in general ask that authors not alter the latex stylesheet so that we can keep a consistent format across papers.
I just want to be clear. I should *not* include section numbers. Is that correct?
 
> > The first point is probably an unfair criterion in that the paper describes a fairly significant data engineering problem and its pragmatic solution
> 
> I haven't had a chance to look through the paper yet, but the 'reproducibility' thing is problematic for two reasons:
> 
> 1. Many authors submit papers from government offices, which are unable to publish code for security reasons; and,
> 2. Many authors submit papers about implementations (as opposed to experiments), which are not reproducible in the classic sense of recovering a given output.
> 
> With that context, please take this review criterion to mean that we encourage openness where possible, but we do not want to require it (and exclude amazing work done by, for example, the US Airforce).

This is an implementation paper rather than an experiment. @dhavide if it is OK with you, I'll leave this part alone, but I will insure that I get some source available at the link provided in the document as soon as I clear it with my security guy. It will likely trickle in.

> > Literature is scant in this area
> 
> It might be helpful to reach out to @choldgraf or one of the other members of the Jupyter core team for some feedback here. They are very nice and helpful, but like all nice and helpful people, they can get a bit oversubscribed.

I've worked with them before and even contributed to From Zero to JupyterHub with Kubernetes, the assortment of "literature" in this area is mostly in the form of `github` issues and stackoverflow queries, not even much in the form of blog posts. I have cited references where I can. But it might be worth a quick pop into their gitter channel.

@deniederhut Question on 
> 5. Please number and add a short description to code listings
Perhaps I misunderstood @gcdeshpande's comment, but I interpret this to mean numbering and captioning in the same sense as I would do on a figure.  While I like the idea, it's not presented that way in the sample papers and I'm not sure how you can do it in the RST file.

**Comment by deniederhut** [comment]
> But it might be worth a quick pop into their gitter channel.

If you don't mind, yes please. One of the great things about Jupyter is the pace of their feature output, but this means that papers will tend to lag behind by 1+ years. We might not be able to recover paper-ish sources for every assertion in your submission, but where possible we want to give proper attribution.

**Comment by deniederhut** [comment]
As for the code sample numbers, I might try using the `:label:` directive.

**Comment by summerswallow-whi** [comment]
> As for the code sample numbers, I might try using the `:label:` directive.

I tried that, doesn't work I don't think the .. code:: class allows captioning or numbering. I know it can be done with custom LaTex, but I noted that the instructions wanted to avoid that.

Also my question in response to your comment got inadvertently buried in a quote.
>Yes, in the past we have not numbered manuscript sections, and in general ask that authors not alter the latex stylesheet so that we can keep a consistent format across papers.

I just want to be clear. I should not include section numbers. Is that correct?

**Comment by deniederhut** [comment]
> I just want to be clear. I should not include section numbers. Is that correct?

Correct.

**Comment by summerswallow-whi** [comment]
As I suspected, there isn't a lot out there. The response I got from the Jupyter community is that they are still working on a blog: https://discourse.jupyter.org/t/co-author-a-blog-post-on-z2jh-security-with-me/2404 But it seems any documentation is fragmented and sparse. I'm not sure a formal literature review is appropriate.

**Comment by dhavide** [comment]
> This is an implementation paper rather than an experiment. @dhavide if it is OK with you, I'll leave this part alone, but I will insure that I get some source available at the link provided in the document as soon as I clear it with my security guy. It will likely trickle in.

This is reasonable. As @deniederhut has clarified the spirit of the reproducibility requirement, I think the description of the implementation in the current manuscript is consistent with that spirit. When you have approval and the link to the source code comes live, that should suffice.

**Comment by summerswallow-whi** [comment]
@dhavide I spoke with our security guy. What we can do is give you (and any other reviewer who desires it) access to the private mirror github repo so that I can populate the code segments quickly. That way you can have access quickly (plus I would appreciate comments to make the supplemental material more navigable)  This will allow my security guy time to vet the material while mitigating potential exposure.

**Comment by choldgraf** [comment]
hmmmm, I'm not sure whether or not there has been a write-up about this. Perhaps @rcthomas or @rpwagner would know if there are good citable documents about JupyterHub security?

**Comment by rpwagner** [comment]
Most of the current material around security in Jupyter is in the form of a few blogs posts and documentation in the Jupyter Project. This makes is a good area to write up ideas and share them, even implementations such as this one. If it helps, there are the [slides from last year's NSF Cybersecurity Workshop](https://scholarworks.iu.edu/dspace/handle/2022/24712) which cover general security in Jupyter and some recommended practices for JupyterHub.

**Comment by rcthomas** [comment]
No, it's kind of an open area as @rpwagner points out.  You might check out whether @tgmachina perhaps has a Livermore TR on the end-to-end-SSL?

**Comment by summerswallow-whi** [comment]
@choldgraf, @rpwagner, @rcthomas. First of all thanks for your input on this matter. This came up since one of the reviewer thought I should have a "literature review" section. I think the perspective is the classic paper approach with a literature review up first. I elected to add a background section instead which does cite some references, but given the current state of Jupyter security literature, I don't know if a formal literature review section makes sense. 

 That said I'm hoping with that said. That this paper might serve as an implementation guide to a secure Jupyter environment on AWS.

**Comment by rpwagner** [comment]
@summerswallow-whi I would suggest a literature review that is somewhat minimal, but broader than just security within the Jupyter Project or JupyterHub. First, you can point out the lack of references, which highlight a need for examples like yours. Then, I recommend some references to how other applications with similar architectures have handled secure deployments. Some of these include providing secure SSH access, which is similar in risk to the interactive notebook interface.

**Comment by dhavide** [comment]
> @dhavide I spoke with our security guy. What we can do is give you (and any other reviewer who desires it) access to the private mirror github repo so that I can populate the code segments quickly. That way you can have access quickly (plus I would appreciate comments to make the supplemental material more navigable) This will allow my security guy time to vet the material while mitigating potential exposure.

@summerswallow-whi   Please don't bother doing this if the code isn't going to be available to the general reader. I don't require access for the purpose of review (the code snippets you've provided give me enough context to figure out what's going on). The concern about reproducibility is nice in an ideal world, but obviously it is impractical in this instance. Oh, and you probably want to remove references to that repo in the final version of the paper given that it's not going to be accessible.

**Comment by summerswallow-whi** [comment]
> @summerswallow-whi Please don't bother doing this if the code isn't going to be available to the general reader. I don't require access for the purpose of review (the code snippets you've provided give me enough context to figure out what's going on). The concern about reproducibility is nice in an ideal world, but obviously it is impractical in this instance. Oh, and you probably want to remove references to that repo in the final version of the paper given that it's not going to be accessible.

@dhavide  I do plan to make it available. In fact, we would like to provide code to help facilitate others who have need. So reproducibility is important in the sense that others can build a system to suit their needs from this work. I was suggesting that I give you access to the staging area if it would aid in your review.

**Comment by summerswallow-whi** [comment]
@dhavide @gcdeshpande Is there any further action I should take at this point?

**Comment by dhavide** [comment]
> @dhavide @gcdeshpande Is there any further action I should take at this point?

With the link correction, I can see I have access to a github repo now (but I assume it's a placeholder and the code will be forthcoming).

I would say at this point, it's mostly minor grammar/style corrections in portions of the manuscript itself. I can take another pass in a few days.

**Comment by summerswallow-whi** [comment]
> With the link correction, I can see I have access to a github repo now (but I assume it's a placeholder and the code will be forthcoming).
> 
> I would say at this point, it's mostly minor grammar/style corrections in portions of the manuscript itself. I can take another pass in a few days.

Thanks. I will re-read for grammar (editing in rst with latex mixed in is a bitch) if you or maybe @deniederhut know of any open source tools that can read the rst without flagging the tons of meta information, that could save us all a lot of work.

**Comment by deniederhut** [comment]
Hey @summerswallow-whi  ! Thanks to some awesome behind the scenes work by @stargaser, we now have experimental support for publishing papers with ORCIDs. If you have one and would like to add it to your paper, it's as simple as adding an `:orcid:` tag under your name in the paper header. [You can see an example of how to do this in the `00_vanderwalt` example paper](https://raw.githubusercontent.com/scipy-conference/scipy_proceedings/2020/papers/00_vanderwalt/00_vanderwalt.rst). If you don't have one, no worries! You can still publish your paper as is.

**Comment by deniederhut** [comment]
Grammar is tricky, but if you are on a unix system there is `aspell` for spelling errors. In one of the other PRs, @ocefpaf recommended `codespell`.

**Comment by gcdeshpande** [comment]
@summerswallow-whi  Thank you for addressing my queries.

**Comment by deniederhut** [comment]
Hi @dhavide ! Are you happy with the updates to this paper?

**Comment by deniederhut** [comment]
Hi @dhavide ! Checking in again - do you feel this paper is ready for inclusion in the proceedings?

**Comment by deniederhut** [comment]
Hi @dhavide ! Today is the deadline for reviewers' final decisions. Do you feel this paper is ready for inclusion in the proceedings?

**Comment by deniederhut** [comment]
Since we haven't been able to reach @dhavide , the proceedings committee has decided to accept this paper on their behalf. A committee member may still reach out with small requested changes about things like a missing DOI in the bibliography, but we are otherwise considering this ready to publish. Congratulations everyone!

**Comment by summerswallow-whi** [comment]
> Since we haven't been able to reach @dhavide , the proceedings committee has decided to accept this paper on their behalf. A committee member may still reach out with small requested changes about things like a missing DOI in the bibliography, but we are otherwise considering this ready to publish. Congratulations everyone!

I'll be standing by today or tomorrow in case there are any requested changes.

**Pull request merged**